### DISCUSSION TITLE
Contour resampling/smoothing/manipulation with Fourier basis

### DISCUSSION THREAD

**Discussion by anttimc** [bfieldtools/bfieldtools#43] [Ideas]
Hi, 

I tested a few months ago an idea of representing contours as continuous functions using a sort of fourier basis. I accidentally added the file in the recently merged PR #39. The test can be found in the master branch here:
https://github.com/bfieldtools/bfieldtools/blob/master/tests/test_contour_basis.py

@rzetter @jiivana could you please try out the test. Could this idea be expanded further?

The file should at least be cleaned up a bit since I did not intend to merge the change as it was.

**Comment by anttimc** [comment]
The underlying idea is that the contours, which are currencly represented as closed polylines, would instead be represented with mapping from unit circle to a curve in **R**<sup>3</sub>. 

In the example/test, the polyline is first thought as this sort of mapping itself. The lenght of the polyline is calculated from the line edge lenghts and normalized to 2π. This gives a one-to-one correspondence between unit circle angle [0, 2π)  and the polyline node coordinates (each coordinate can be thought as a separate function of the angle). After this, the Fourier basis matrix is calculated for angle values at the nodes and the basis function weights are fitted to match the sampled coordinate functions.

This gives a continuous representation of the contours, which can for example be resampled as wished. Smoothing effect can be accomplished by penalizing for high frequencies in the fitting as I did in the example.

**Reply by anttimc** [reply]
I'm writing a small latex memo about this to explain the code a bit better and to understand certain things better myself (some things have just worked out with intuition, trial and error). I think I can simplify my local changes a bit before pushing

**Comment by anttimc** [comment]
Here are the results of  an extended example that I have not pushed yet:

Countour smoothing and resampling:
![coil_fourier_interpolated](https://user-images.githubusercontent.com/38734060/148202955-b8bae78e-c552-44f4-8629-7241c173f743.png)
The dashed lines and points '*' are the result of standard contouring on a low quality mesh. The solid lines are curves fitted using the Fourier basis.
 
Merging the curves to one contiuous coil shape:
![coil_fourier_merged](https://user-images.githubusercontent.com/38734060/148202968-2e27941d-57ef-4f07-bc6d-3726504db069.png)
 The curves represented as Fourier coefficients are merged by interpolating in the coefficient space with a Bezier-like method. Bezier-curves are constructed by interpolating the cartesian coordinates of the nodal points, but in this case I interpolated between the different contours in the Fourier coefficient space with some tweaks in the interpolation weights.

**Reply by jiivana** [reply]
Super interesting. This example looks very promising. I would love to know the details. Can you share the latex memo when it is semi-ready? 

I think I understood roughly what is going on but how do you know where to join the different loops?

**Reply by anttimc** [reply]
>I think I understood roughly what is going on but how do you know where to join the different loops?

Case-by-case you have to make a rule where to start the parametrization of the different loops. The starting points need be aligned. Generally, the starting points would aligned on a line or curve between the extrema of the stream function.

**Comment by rzetter** [comment]
Very nice! 

I tested the functionality in test_contour_basis.py out with a couple more complex coils, it seems to work well! Some tweaking of Nfreqs and lambda is obviously needed on a case-by-case basis. The ability to create these continuous contours will alleviate the need for as high-resolution meshes for the stream-function, but perhaps not entirely. We still need a good enough representation of the continuous surface current. In any case, I think this is very nice additional functionality. What do you think, should it be added LineConductor or added as a separate new class?

The merging of these independent loops into a single continuous conductor looks very promising! I'd love to how it's done, at the moment I don't have a good intuition for how it is done. E.g. how are the nodes between which to interpolate chosen?

Below is a 3D snapshot from [this example](https://bfieldtools.github.io/auto_examples/coil_design/spherical_harmonics_coil_design.html#sphx-glr-auto-examples-coil-design-spherical-harmonics-coil-design-py) using the same rough 10x10 plane mesh (original polyline in black, Fourier-basis representation in red):
![image](https://user-images.githubusercontent.com/39831259/148208165-b7a41e8f-2f9e-4397-8212-9101f33565d5.png)

**Reply by anttimc** [reply]
> We still need a good enough representation of the continuous surface current. 

yes, I agree. The contouring needs certain resolution, but the irregularities in the contours due to discretization could be smoothed out using this technique. One main benefit in the new representation is that the contuors can be resampled with  regular spacing.

> What do you think, should it be added LineConductor or added as a separate new class?

Maybe a separate class? It could take a LineConductor as input to generate the Fourier coefficients. It could also output a LineConductor for further calculations (e.g. fields).

**Comment by anttimc** [comment]
Hello, I think I figured out something even better for merging. After resampling the contours evenly I concatenated the coordinates of the contours. Once for the frontside of a PCB and a second time in reverse order for the backside. Then I low-pass-filtered the coordinate functions using FFT and this is the result:
![kuva](https://user-images.githubusercontent.com/38734060/151713684-5e9f7fc8-9241-49dc-8dfe-bf9b1530231e.png)

The two colors would go on the two sides of the board bringing the starting point and end point of the conductor close together.

**Reply by anttimc** [reply]
@rzetter @jiivana this is probably a better/simpler/cleaner approach than what I already wrote about in the memo. What do you think?

**Reply by jiivana** [reply]
It looks nice and it is cool that it is automatic. I am a bit worried about how much that merging changes the generated magnetic field?

**Reply by rzetter** [reply]
The result looks great, that looks like a lot less work. Do you still have to manually decide at what node indices to concatenate the loops at? I'd love to try validating the functionality with some more complex coil topologies.

**Reply by anttimc** [reply]
I pushed the example functions to a new branch, contour_merge: https://github.com/bfieldtools/bfieldtools/commit/bce545bbac50cfdb950d48a3ce1e2570cc3c9a4c

The example file is now located in misc_examples, instead of tests.

@rzetter if you can generalize this for a more complicated stream-function pattern and surface geometry. We could try to start to refactor the code in a more general format.

> Do you still have to manually decide at what node indices to concatenate the loops at?

Yes, but not necessarily manually. Take a look at "zero_phase_func" in the example. It takes the polyline in and returns the index of the polyline node which is aligned with zero-phase in the Fourier representation. The rest is then automatic. I think the phase could be fine tuned even better.

**Reply by MengerSponge** [reply]
> It looks nice and it is cool that it is automatic. I am a bit worried about how much that merging changes the generated magnetic field?

I'm just getting up to speed, but check out this paper from Crawford and Straley: https://aip.scitation.org/doi/full/10.1063/5.0063057

Quantifying the field perturbation introduced with the resampling/smoothing would make this a *super* useful technique!

**Reply by anttimc** [reply]
I don't have institutional access anymore. Is there a preprint or something available? Sounds pretty interesting.

**Reply by MengerSponge** [reply]
[Crawford and Straley - 2021 - Improved prescription for winding an electromagnet.pdf](https://github.com/bfieldtools/bfieldtools/files/8006145/Crawford.and.Straley.-.2021.-.Improved.prescription.for.winding.an.electromagnet.pdf)
[Crawford - 2021 - The physical meaning of the magnetic scalar potent.pdf](https://github.com/bfieldtools/bfieldtools/files/8006146/Crawford.-.2021.-.The.physical.meaning.of.the.magnetic.scalar.potent.pdf)

**Reply by anttimc** [reply]
Thank you, the paper has some very nice ideas (some of those are also covered for example in our bfieldtools Part I paper and realized in our examples). 

One of the new ideas to me is that in a axial symmetrical case the field of the "jumper wires" would produce zero field inside. The paper also discusses the more general case, and this made me to rethink this particular case of doubled sided current in this discussion.

If we consider the limit of very narrowly spaced loops, we could think the current that joins the loops as it's own current density component J1 (as in the paper). As the wires on the other side of the board come back to the point of origin symmetrically, the current density component joining the loops on the backside J2 would be exactly the opposite of J1, i.e., J2=-J1. Therefore, there would be no contribution from joining the loops to the field outside the board (in the limit). This should also answer @jiivana 's question earlier in this thread.

**Reply by jiivana** [reply]
Those are really interesting papers!

I understand what you mean but I would like to calculate the magnetic field of the joined loops and compare that to the field by the original loops and add that calculation to this example. Just to be sure, you know. If you implement this coil in practice there will be a small offset between the current layers. In most cases, the offset will be negligible but in some cases it might be significant enough to cause minor deviations from the desired field patterns.

**Comment by anttimc** [comment]
I forked our repo, and try to more cleaner/more useful code out of these examples. A year ago, I also had an idea about generalizing the idea for stream functions with more than two extrema. Let's see what I can make, the aim is to make an PR once, I have some polished code.
